velox/exec/QuerySplitTracer.cpp

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/exec/QuerySplitTracer.h"
+#include "QueryTraceUtil.h"
+#include "velox/connectors/hive/HiveConnectorSplit.h"
+#include "velox/exec/QueryTraceTraits.h"
+
+using namespace facebook::velox::connector::hive;
+
+namespace facebook::velox::exec::trace {
+/// Used to record and load the input splits from a tracing 'TableScan'
+/// operator, and for getting the traced splits when relaying 'TableScan'.
+///
+/// Currently, it only works with 'HiveConnectorSplit'. In the future, it will
+/// be extended to handle more types of splits, such as
+/// 'IcebergHiveConnectorSplit'.
+QuerySplitTracer::QuerySplitTracer(std::string traceDir)
+    : traceDir_(std::move(traceDir)),
+      fs_(filesystems::getFileSystem(traceDir_, nullptr)) {
+  VELOX_CHECK_NOT_NULL(fs_);
+}
+
+void QuerySplitTracer::write(const exec::Split& split) {
+  VELOX_CHECK(!split.hasGroup());
+  VELOX_CHECK(split.hasConnectorSplit());
+  const auto splitInfoFile = fs_->openFileForWrite(fmt::format(
```

Comment:
Are you going to have a shared split writer? Then you can have a lock inside and persist the split info and with the first uint32_t as length, serialized data and the checksum to detect corruption. Then on replay, you can recover from that even in case of query failure? Thanks!